### PR TITLE
chore: manifest migration to V3

### PR DESCRIPTION
**Pull request by DragosIorgulescu**
## Changes

- updated `manifest.json` to match the new V3 API - [DR spec](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/618987566/WT-1764+Manifest+V3#Updating-Manifest)
- migrate to service workers - [DR spec](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/618987566/WT-1764+Manifest+V3#Service-Worker-Migration)
  - works out of the box with the current implementation
- `browserAction` deprecation in favor of `action` - [DR spec](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/618987566/WT-1764+Manifest+V3#Action-API)
- validate other API changes - [DR spec](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/618987566/WT-1764+Manifest+V3#Deprecated-APIs)
  - no changes needed, we are not using any of these
- remove `webextension-polyfill-ts` and use the main `webextension-polyfill` lib w/ included types
- extension security checklist: [DR spec](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/618987566/WT-1764+Manifest+V3#Extension-Security)
  - remotely hosted code:  n/a
  - remote executions: n/a
  - content security policy: we are no longer pulling `script-src 'self' https://www.google-analytics.com` so this is no longer an issue
  - cross cookies in service worker: working as expected
 

## Events

- [x] validate events are firing 

### On those affected packages:
- ~~Have you done sanity checks in the webapp?~~
- [x] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-1907 #done

### CODE REVIEW

File: packages/extension/src/manifest.json [hunk 1]
```diff
@@ -10,30 +10,38 @@
...
-  "__dev__permissions": ["storage", "http://localhost/", "http://*.local.com/"],
-  "optional_permissions": ["topSites", "*://*/*"],
-  "content_security_policy": "script-src 'self' https://www.google-analytics.com; object-src 'self'",
-  "__firefox__content_security_policy": "object-src 'self'",
-  "browser_action": {
+  "__dev__permissions": ["storage"],
+  "__dev__host_permissions": ["http://localhost/", "http://*.local.com/"],
+  "optional_permissions": ["topSites"],
+  "optional_host_permissions": [ "*://*/*"],
+  "content_security_policy": {
```
Comment (idoshamun) [inline packages/extension/src/manifest.json:23]: I see we removed the specific definition for firefox, is it on purpose?

**Review by idoshamun — COMMENTED** [review]
Can't believe v3 is coming together finally! Thanks a lot

**Review by rebelchris — COMMENTED** [review]
Oh wow, expected we really needed more changes in the background worker to be honest 🤯 
Let's move to testing phase.

I do also support Ido's question, was that a intentional removal?

File: packages/extension/src/background/index.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import 'content-scripts-register-polyfill';
-import { browser, Runtime, Tabs } from 'webextension-polyfill-ts';
+import browser, { runtime, Runtime, Tabs } from 'webextension-polyfill';
```
Comment (capJavert) [inline packages/extension/src/background/index.ts:2]: Just use `browser.runtime`? It will be easier to refactor down the line if we change the polyfill since `browser` is usually available as a global in extension context.

File: packages/shared/src/components/tooltips/ChangelogTooltip.tsx [hunk 3]
```diff
@@ -56,8 +56,8 @@ function ChangelogTooltip({
           );
         }
 
-        const browser = await import('webextension-polyfill-ts').then(
-          (mod) => mod.browser,
+        const browser = await import('webextension-polyfill').then(
```
Comment (capJavert) [inline packages/shared/src/components/tooltips/ChangelogTooltip.tsx:59]: Nice, and now we have type support for it 👌

**Review by capJavert — COMMENTED** [review]
Nice work, at this point I have no major comments, really smooth implementation and I like the fact that aside from mocks you did not have to update any tests and are all 🟢

File: packages/extension/src/manifest.json (hunk 1 above)
Comment (DragosIorgulescu) [reply packages/extension/src/manifest.json:23]: good catch, since we removed google analytics, there was no meaningful diff between the regular content security policy and the firefox specific one. Was there a particular reason `script-src 'self'` was not initially included there?

We can see during testing in firefox if this causes any problems or triggers a permissions reset.

File: packages/extension/src/background/index.ts (hunk 2 above)
Comment (capJavert) [reply packages/extension/src/background/index.ts:2]: I see you use `browser` top level in all other places so this is the only different case to adjust.

File: packages/extension/src/manifest.json (hunk 1 above)
Comment (idoshamun) [reply packages/extension/src/manifest.json:23]: Fair enough, I'm not sure why we added it in the first place but we can leave it for testing

File: packages/extension/src/background/index.ts (hunk 2 above)
Comment (DragosIorgulescu) [reply packages/extension/src/background/index.ts]: good catch, updated 🙏

File: packages/extension/src/manifest.json (hunk 1 above)
Comment (DragosIorgulescu) [reply packages/extension/src/manifest.json:23]: having this in also ensures protection agains cross site scripting and other things, so it's a good default to have
 
BTW, I found the answer from daildev search here very nice: https://app.daily.dev/search?id=f69e7f9a-7e23-11ee-9cee-42010a007e0c

File: packages/extension/src/background/index.ts [hunk 4]
```diff
@@ -12,6 +11,8 @@ import { BOOT_LOCAL_KEY } from '@dailydotdev/shared/src/contexts/common';
 import { ExtensionMessageType } from '@dailydotdev/shared/src/lib/extension';
 import { getContentScriptPermissionAndRegister } from '../lib/extensionScripts';
 
+const client = new GraphQLClient(graphqlUrl, { fetch: globalThis.fetch });
```
Comment (sshanzel) [inline packages/extension/src/background/index.ts:14]: Very neat solution. Have you checked whether we can completely avoid having to go through the process below for regular GraphQL queries? Meaning, without using the proxy for the `request` method and the queries would still work?

Comment (DragosIorgulescu) [reply packages/extension/src/background/index.ts:14]: @sshanzel good point, this is only required in the background service worker due to its restricted access to various APIs (such as the `XMLHttpRequest` object). Tested it as well, of course, this works as expected outside of the background worker 🙏

File: packages/extension/webpack.config.js [hunk 5]
```diff
@@ -77,9 +52,7 @@ module.exports = {
   resolve: {
     extensions: ['.svg', '.ts', '.tsx', '.js', '.json'],
     alias: {
-      'webextension-polyfill-ts': path.resolve(
-        path.join(__dirname, 'node_modules', 'webextension-polyfill-ts'),
-      ),
+      fetch: false, // or point to a file that exports the global fetch
```
Comment (capJavert) [inline packages/extension/webpack.config.js:55]: what is this for, was this culprit for xmlhttprequest usage in compiled bundle?

File: packages/extension/webpack.config.js [hunk 6]
```diff
@@ -191,8 +164,6 @@ module.exports = {
     new CopyWebpackPlugin({
       patterns: [{ from: 'public', to: '.' }],
     }),
-    // plugin to enable browser reloading in development mode
-    extensionReloaderPlugin,
```
Comment (capJavert) [inline packages/extension/webpack.config.js:195]: We should open the backlog task to fix reloader down the line.

File: packages/extension/src/lib/extensionScripts.ts [hunk 7]
```diff
@@ -1,29 +1,37 @@
...
+let hasInjectedScripts = false;
+
+export const registerBrowserContentScripts = async (): Promise<void> => {
+  if (hasInjectedScripts) {
+    return null;
+  }
+
+  await browser.scripting.registerContentScripts([
+    {
+      id: `companion-${Date.now()}`,
```
Comment (capJavert) [inline packages/extension/src/lib/extensionScripts.ts:18]: Is there a better way to do this? Worried we will register a bunch of same content scripts and load them to user browser. I suspect this might be due to content script being registered in each tab/context our extension and companion are loaded.

File: packages/extension/webpack.config.js [hunk 8]
```diff
@@ -6,7 +6,6 @@ const TerserPlugin = require('terser-webpack-plugin');
 const CopyWebpackPlugin = require('copy-webpack-plugin');
 const HtmlWebpackPlugin = require('html-webpack-plugin');
 const { CleanWebpackPlugin } = require('clean-webpack-plugin');
-const ExtensionReloader = require('webpack-ext-reloader');
```
Comment (capJavert) [inline packages/extension/webpack.config.js:9]: we can also remove from package.json?

**Review by capJavert — CHANGES_REQUESTED** [review]
Almost there, some notes from latest changes.

File: packages/extension/src/lib/extensionScripts.ts (hunk 7 above)
Comment (capJavert) [reply packages/extension/src/lib/extensionScripts.ts:18]: This looks to me as something we need to solve before production.

Comment (DragosIorgulescu) [reply packages/extension/src/lib/extensionScripts.ts:25]: good point @capJavert I'll do some digging into this one

File: packages/extension/src/background/index.ts (hunk 4 above)
Comment (sshanzel) [reply packages/extension/src/background/index.ts:15]: Nice, thanks for the verification. The reason I was curious is, that this will remove the overhead of using the `useBackgroundRequest` which complicates the process when making a query/mutation. This means, the requests would automatically work without using those hooks. I understand this is only to migrate the Manifest, so I think we should create a ticket for it to migrate the implementation. Would also make new implementations work automatically as opposed to introducing the said hook every time we do something. cc @rebelchris

File: packages/extension/src/lib/extensionScripts.ts (hunk 7 above)
Comment (capJavert) [reply packages/extension/src/lib/extensionScripts.ts:32]: @DragosIorgulescu so I was just checking something by removing `Date.now()` above. To my surprise I could not get the error we had before. Maybe it was fixed in other way? So yeah, maybe just remove it then?

File: packages/storybook/stories/Button.stories.tsx [hunk 9]
```diff
@@ -1,25 +1,12 @@
-import type { Meta, StoryObj } from '@storybook/react';
-import classNames from 'classnames';
-
-import {
-  Button,
-  ButtonSize,
-} from '@dailydotdev/shared/src/components/buttons/Button';
-import ShareIcon from '@dailydotdev/shared/src/components/icons/Share';
+import type { Meta, StoryObj } from "@storybook/react";
```
Comment (rebelchris) [inline packages/storybook/stories/Button.stories.tsx:1]: Why are these changes needed?

File: packages/extension/src/lib/extensionScripts.ts (hunk 7 above)
Comment (idoshamun) [reply packages/extension/src/lib/extensionScripts.ts:32]: Agree this cannot remain as is before publishing. It's too risky

File: packages/storybook/stories/Button.stories.tsx (hunk 9 above)
Comment (DragosIorgulescu) [reply packages/storybook/stories/Button.stories.tsx]: good point, not sure, maybe a bad merge conflict. Removed 👍

File: packages/extension/src/lib/extensionScripts.ts (hunk 7 above)
Comment (DragosIorgulescu) [reply packages/extension/src/lib/extensionScripts.ts]: this should be properly handled now 👍

File: packages/extension/webpack.config.js (hunk 6 above)
Comment (DragosIorgulescu) [reply packages/extension/webpack.config.js:195]: ticket here: https://dailydotdev.atlassian.net/browse/MI-107

File: packages/extension/src/background/index.ts (hunk 4 above)
Comment (sshanzel) [reply packages/extension/src/background/index.ts:15]: @DragosIorgulescu we should maybe add a ticket for this. To make implementations less complicated.

File: packages/extension/src/manifest.json [hunk 10]
```diff
@@ -10,30 +11,56 @@
...
     "https://dailynow.co/",
     "https://*.dailynow.co/"
   ],
-  "__dev__permissions": ["storage", "http://localhost/", "http://*.local.com/"],
-  "optional_permissions": ["topSites", "*://*/*"],
-  "content_security_policy": "script-src 'self' https://www.google-analytics.com; object-src 'self'",
-  "__firefox__content_security_policy": "object-src 'self'",
-  "browser_action": {
+  "__firefox|prod__permissions": [
+    "scripting",
```
Comment (rebelchris) [inline packages/extension/src/manifest.json:22]: This is a additional request right?
Why do we need it for firefox (and chrome for that matter)

File: packages/shared/src/components/ExtensionPermissionsPrompt.tsx [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import React, { ReactElement, useEffect } from 'react';
+import Logo, { LogoPosition } from './Logo';
+import { OnboardingTitleGradient } from './onboarding/common';
+import { Button, ButtonSize } from './buttons/Button';
```
Comment (rebelchris) [inline packages/shared/src/components/ExtensionPermissionsPrompt.tsx:4]: This should be buttonV2

File: packages/shared/src/components/ExtensionPermissionsPrompt.tsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+  const handleRequestHostPermissions = async () => {
+    const success = await requestHostPermissions({ origins });
+    if (success) {
+      refetchHostPermissions();
+    }
+  };
+
+  return (
+    <div className="flex max-h-[100vh] min-h-[100vh] flex-col items-center justify-center overflow-hidden px-7 text-center antialiased">
```
Comment (rebelchris) [inline packages/shared/src/components/ExtensionPermissionsPrompt.tsx:30]: Why don't we use `max-h-screen` here?

File: packages/shared/src/components/post/PostComments.tsx [hunk 13]
```diff
@@ -58,7 +59,7 @@ export function PostComments({
         requestMethod(
           graphqlUrl,
           POST_COMMENTS_QUERY,
-          { postId: id, [initialDataKey]: comments, first: 500 },
+          { postId: id, [initialDataKey]: comments, first: 50 },
```
Comment (rebelchris) [inline packages/shared/src/components/post/PostComments.tsx:62]: Not against it, but why did we change this here?

File: packages/shared/src/hooks/useHostPermissionStatus.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { useQuery } from '@tanstack/react-query';
+import { disabledRefetch } from '../lib/func';
+import { useExtensionContext } from '../contexts/ExtensionContext';
+
+export const hostKey = ['host_key'];
```
Comment (rebelchris) [inline packages/shared/src/hooks/useHostPermissionStatus.ts:5]: Why don't we use `generateQueryKey` here?

File: packages/shared/src/lib/constants.ts [hunk 15]
```diff
@@ -59,3 +59,8 @@ export const authUrl =
 export const heimdallUrl = isDevelopment
   ? process.env.NEXT_PUBLIC_HEIMDALL_URL || 'http://127.0.0.1:3000'
   : authUrl;
+
+export const extensionHostPermissions = [
+  'https://daily.dev/',
+  'https://*.daily.dev/',
+];
```
Comment (rebelchris) [inline packages/shared/src/lib/constants.ts:66]: Unused?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Some minor comments :)

File: packages/shared/src/components/post/PostComments.tsx (hunk 13 above)
Comment (sshanzel) [reply packages/shared/src/components/post/PostComments.tsx:62]: Our post comments are not yet paginated, this would not show comments beyond 50.

File: packages/extension/webpack.config.js (hunk 8 above)
Comment (DragosIorgulescu) [reply packages/extension/webpack.config.js:9]: I see no harm in keeping it since we are planning on adding it back pretty soon after releasing this anyway

File: packages/extension/src/manifest.json (hunk 10 above)
Comment (DragosIorgulescu) [reply packages/extension/src/manifest.json:22]: This is needed so we can make the request to register the companion content script: https://github.com/dailydotdev/apps/pull/2332/files/d9b5d2c91912ca948c32ea17e0a4532d336e3fce#diff-9d3bc2384cfa13933a9b00959c4da1058f54cc6338b656556dab709bc868ba5eR42

this has changed from `browser.contentScripts.register` which was used before and is no longer supported in V3.

Without this, we get the following error:
```
TypeError: D().scripting is undefined
```

We could use different APIs, but if this works on V2 as well, not sure if it's worth it to have an `if/else` for this

Comment (capJavert) [reply packages/extension/src/manifest.json:22]: It does not trigger any permission popups?

File: packages/extension/webpack.config.js (hunk 8 above)
Comment (capJavert) [reply packages/extension/webpack.config.js:9]: Fine either way, add a TODO maybe or note so we don't forget what it is.

File: packages/extension/src/manifest.json (hunk 10 above)
Comment (DragosIorgulescu) [reply packages/extension/src/manifest.json:22]: > It does not trigger any permission popups?

No new permission pop-ups are generated by this. We can validate this again through a test release of the firefox extension

File: packages/extension/webpack.config.js (hunk 8 above)
Comment (DragosIorgulescu) [reply packages/extension/webpack.config.js:9]: I added a ticket for this here: https://dailydotdev.atlassian.net/browse/MI-107

File: packages/extension/src/manifest.json (hunk 10 above)
Comment (capJavert) [reply packages/extension/src/manifest.json:22]: Fine by me.

File: packages/extension/src/manifest.json [hunk 16]
```diff
@@ -10,30 +11,56 @@
     "48": "icons/48.png",
     "128": "icons/128.png"
   },
-  "__prod__permissions": [
+  "__prod__permissions": ["scripting"],
```
Comment (idoshamun) [inline packages/extension/src/manifest.json:14]: Is this needed?

File: packages/extension/src/lib/extensionScripts.ts [hunk 17]
```diff
@@ -1,29 +1,63 @@
...
+export const registerBrowserContentScripts = async (): Promise<void> => {
+  if (hasInjectedScripts) {
+    return null;
+  }
+
+  const registeredScripts = await browser.scripting.getRegisteredContentScripts(
+    {
+      ids: [companionScriptId],
+    },
+  );
```
Comment (capJavert) [inline packages/extension/src/lib/extensionScripts.ts:35]: I think this can just always call `browser.scripting.getRegisteredContentScripts` because if for any reason user disallows content script from chrome menu `hasInjectedScripts` will be outdated. I don't think actual call to `getRegisteredContentScripts` will be much of a overhead because it is a Promise to local browser APIs. 

So basically I would remove `hasInjectedScripts` optimization.

**Review by capJavert — APPROVED** [review]
I don't see anything blocking anymore, I would only simplify loading of content script even more to avoid edge cases. Approving either way.

File: packages/extension/src/lib/extensionScripts.ts (hunk 17 above)
Comment (capJavert) [reply packages/extension/src/lib/extensionScripts.ts]: You can remove `hasInjectedScripts` completely now right, meaning declaration and all assignments?

Comment (DragosIorgulescu) [reply packages/extension/src/lib/extensionScripts.ts]: Since we are still using V2 API on FF via `browser.contentScripts.register`, we still need it for now

Comment (capJavert) [reply packages/extension/src/lib/extensionScripts.ts]: You are right, tnx for clarification 🚀

File: packages/shared/src/hooks/useExtensionPermission.ts [hunk 18]
```diff
@@ -1,49 +0,0 @@
-import { QueryClient, useQueryClient } from '@tanstack/react-query';
```
Comment (sshanzel) [inline packages/shared/src/hooks/useExtensionPermission.ts:1]: Is this file deleted already? Looks like the content was just removed but not the file entirely.

**Review by sshanzel — APPROVED** [review]
Nothing else to add from my end. Definitely a great task. This was no easy task and now we are ready for release 🚀

File: packages/shared/src/hooks/useExtensionPermission.ts (hunk 18 above)
Comment (DragosIorgulescu) [reply packages/shared/src/hooks/useExtensionPermission.ts:1]: great catch, not sure how this happened, maybe some merge conflict

**Review by rebelchris — APPROVED** [review]
Nothing left from my side code wise.
